beacon-chain/light-client/store.go

```diff
@@ -0,0 +1,312 @@
+		}
+	} else {
+		if updateSignaturePeriod != storePeriod {
+			return errors.New("update skips a sync committee period")
+		}
+	}
+
+	// Verify update is relevant
+	updateAttestedPeriod := computeSyncCommitteePeriodAtSlot(update.GetAttestedHeader().Slot)
+	updateHasNextSyncCommittee := !s.isNextSyncCommitteeKnown() && (update.isSyncCommiteeUpdate() && updateAttestedPeriod == storePeriod)
+	if !(update.GetAttestedHeader().Slot > s.FinalizedHeader.Slot || updateHasNextSyncCommittee) {
+		return errors.New("update is not relevant")
+	}
+
+	// Verify that the finality branch, if present, confirms finalized header to match the finalized checkpoint root
+	// saved in the state of attested header. Note that the genesis finalized checkpoint root is represented as a zero
+	// hash.
+	if !update.isFinalityUpdate() {
+		if update.GetFinalizedHeader() != &(ethpbv1.BeaconBlockHeader{}) {
+			return errors.New("finality branch is present but update is not finality")
+		}
+	} else {
+		var finalizedRoot [32]byte
+		if update.GetFinalizedHeader().Slot == genesisSlot {
+			if update.GetFinalizedHeader() != &(ethpbv1.BeaconBlockHeader{}) {
+				return errors.New("finality branch is present but update is not finality")
+			}
+			finalizedRoot = [32]byte{}
+		} else {
+			var err error
+			if finalizedRoot, err = update.GetFinalizedHeader().HashTreeRoot(); err != nil {
+				return err
+			}
+		}
+		if !trie.VerifyMerkleProofWithDepth(
+			update.GetAttestedHeader().StateRoot,
+			finalizedRoot[:],
+			getSubtreeIndex(ethpbv2.FinalizedRootIndex),
+			update.GetFinalityBranch(),
+			uint64(ethpbv2.FloorLog2(ethpbv2.FinalizedRootIndex))) {
+			return errors.New("finality branch is invalid")
+		}
+	}
+
+	// Verify that the next sync committee, if present, actually is the next sync committee saved in the state of the
+	// attested header
+	if !update.isSyncCommiteeUpdate() {
+		if update.GetNextSyncCommittee() != &(ethpbv2.SyncCommittee{}) {
+			return errors.New("sync committee branch is present but update is not sync committee")
+		}
+	} else {
+		if updateAttestedPeriod == storePeriod && s.isNextSyncCommitteeKnown() {
+			if update.GetNextSyncCommittee() != s.NextSyncCommittee {
+				return errors.New("next sync committee is not known")
+			}
+		}
+		if !trie.VerifyMerkleProofWithDepth(
+			update.GetAttestedHeader().StateRoot,
+			hashTreeRoot(update.GetNextSyncCommittee())[:],
+			getSubtreeIndex(ethpbv2.NextSyncCommitteeIndex),
+			update.GetNextSyncCommitteeBranch(),
+			uint64(ethpbv2.FloorLog2(ethpbv2.NextSyncCommitteeIndex))) {
+			return errors.New("sync committee branch is invalid")
+		}
+	}
+
+	var syncCommittee *ethpbv2.SyncCommittee
+	// Verify sync committee aggregate signature
+	if updateSignaturePeriod == storePeriod {
+		syncCommittee = s.CurrentSyncCommittee
+	} else {
+		syncCommittee = s.NextSyncCommittee
+	}
+	participantPubkeys := []common.PublicKey{}
+	for i, bit := range syncAggregate.SyncCommitteeBits {
```

Comment:
I'm not sure that's enforced anywhere so maybe play it safe?

```go
type Bitvector512 []byte
```

Reply:
Oh, I see.    They declare this as a slice (might be better to declare it as an array with 64 length)

Since they provide the receiver method `BitAt` for this type, I think we might have to use it.   Otherwise, the `range` loop only loops 64 times, and the `bit` local variable in the `range` loop will be some big number most of the time.